src/science/science.py

```diff
@@ -0,0 +1,263 @@
+'''
+Writes, reads and parses NMEA like messages from the onboard
+science nucleo to operate the science boxes and get relevant data
+'''
+import serial
+# import Adafruit_BBIO.UART as UART  # for beaglebone use only
+import numpy as np
+from enum import Enum
+import rospy
+from mrover import AutonLED, Enable, Heater, Servo, Spectral, Thermistor
+
+
+class ScienceBridge():
+
+    def __init__(self):
+
+        # UART.setup("UART4")  # Specific to beaglebone
+        
+        self.NMEA_HANDLE_MAPPER = {
+            "AUTOSHUTOFF": self.heater_shut_off_handler,
+            "HEATER": self.heater_state_handler,
+            "SPECTRAL": self.spectral_handler,
+            "THERMISTOR": self.thermistor_handler,
+            "TRIAD": self.triad_handler
+        }
+        self.NMEA_PUBLISHER_MAPPER = {
+            "AUTOSHUTOFF": rospy.Publisher('heater_auto_shut_off_data', Enable, queue_size=1),
+            "HEATER": rospy.Publisher('heater_state_data', Heater, queue_size=1),
+            "SPECTRAL": rospy.Publisher('spectral_data', Spectral, queue_size=1),
+            "THERMISTOR": rospy.Publisher('thermistor_data', Thermistor, queue_size=1),
+            "TRIAD": rospy.Publisher('spectral_triad_data', Spectral, queue_size=1)
+        }
+        self.NMEA_MESSAGE_MAPPER = {
+            "AUTOSHUTOFF": Enable(),
+            "HEATER": Heater(),
+            "SPECTRAL": Spectral(),
+            "THERMISTOR": Thermistor(),
+            "TRIAD": Spectral()
+        }
+        self.max_error_count = 20
+        self.sleep = .01
+        self.led_map = {
+            "Red": 0,
+            "Blue": 1,
+            "Green": 2
+            "Off": 3
+        }
+        # Mapping of onboard devices to mosfet devices
+        self.mosfet_dev_map = {
+            "arm_laser" : 1
+            "raman_laser" : 10
+            "uv_bulb" : 1
+            "uv_led" : 4
+            "white_led" : 5
+        }
+        self.heater_map = [7, 8, 9]
+        self.UART_TRANSMIT_MSG_LEN = 30
+
+    def __enter__(self):
+        '''
+        Opens a serial connection to the nucleo
+        '''
+        self.ser = serial.Serial(
+            port='/dev/ttyTHS0',
```

Comment:
we should set up some kind of config file with all the ports for the devices if possible. Maybe this should be part of the launchfile so that this stuff is in the parameter server?